src/server/server.ts

```diff
@@ -0,0 +1,403 @@
+              formatResponse: <any>undefined,
+              formatError: <any>undefined,
+              callback: <any>undefined,
+            };
+            let promisedParams = Promise.resolve(baseParams);
+
+            if (this.onRequest) {
+              promisedParams = Promise.resolve(this.onRequest(parsedMessage, baseParams, connectionContext.socket));
+            }
+
+            // if we already have a subscription with this id, unsubscribe from it first
+            this.unsubscribe(connectionContext, reqId);
+
+            promisedParams.then(params => {
+              if (typeof params !== 'object') {
+                const error = `Invalid params returned from onRequest! return values must be an object!`;
+                this.sendError(connectionContext, reqId, { message: error });
+
+                throw new Error(error);
+              }
+
+              // NOTE: This is a temporary code to identify if the request is a real subscription or only a query/mutation.
+              // As soon as subscription manager starts handling complete function from the observable calling  the
+              // callback function with (null, null) this can be replaced
+              const isSubscription = this.isASubscriptionRequest(params.query, params.operationName);
+
+              // Create a callback
+              // Error could be a runtime exception or an object with errors
+              // Result is a GraphQL ExecutionResult, which has an optional errors property
+              params.callback = (error: any, result: any) => {
+                if (connectionContext.requests[reqId]) {
+                  if (result) {
+                    this.sendMessage(connectionContext, reqId, MessageTypes.GQL_DATA, result);
+                  }
+
+                  if (error) {
+                    const errorsToSend = error.errors ?
+                      { errors: error.errors } :
+                      { errors: [ { message: error.message } ]};
+                    this.sendMessage(connectionContext, reqId, MessageTypes.GQL_DATA, errorsToSend);
+                  }
+                }
+
+                // NOTE: This is a temporary code to identify if the request is a real subscription or only a query/mutation.
+                // As soon as subscription manager starts handling complete function from the observable calling  the
+                // callback function with (null, null) this can be replaced
+                if (!isSubscription) {
+                  this.unsubscribe(connectionContext, reqId);
+                  this.sendMessage(connectionContext, reqId, MessageTypes.GQL_COMPLETE, null);
+                }
+              };
+              return this.subscriptionManager.subscribe(params);
+            }).then((graphqlReqId: number) => {
+              connectionContext.requests[reqId] = graphqlReqId;
+              this.sendMessage(connectionContext, reqId, MessageTypes.SUBSCRIPTION_SUCCESS, undefined);
```

Comment:
please split it under a different `then` block,
then add a comment that it should be removed as soon as the older protocol is deprecated